### PR TITLE
Adds support to create repository from a template

### PR DESCRIPTION
**Pull request by isouza-daitan**
This PR adds a new `create_repo_from_template` method to `AuthenticatedUser` and `Organization` that uses the new `/generate` endpoint of a repo. I followed the same pattern as the similar `create_fork` method;

I also added a new `is_template` attribute to the `Repository` class. This attribute will only be sent by Github when we use the feature's `Accept` header which I only added to the POST to create the new repo. Let me know if I should also add it to `get_repo` calls.

The documentation mentions a template_repository attribute but Github never sends it.

### CODE REVIEW

File: scripts/add_attribute.py [hunk 1]
```diff
@@ -138,4 +138,4 @@
 
 with open(fileName, "wb") as f:
     for line in newLines:
-        f.write(line + "\n")
+        f.write((line + "\n").encode())
```
Comment (isouza-daitan) [inline scripts/add_attribute.py]: change to make this work with python 3

File: tests/AuthenticatedUser.py [hunk 2]
```diff
@@ -659,6 +659,22 @@ def testCreateFork(self):
...
+        self.assertEqual(repo.is_template, False)
+
+    def testCreateRepoFromTemplateWithAllArguments(self):
+        template_repo = self.g.get_repo("actions/hello-world-docker-action")
+
+        description = "My repo from template"
+        private = True
+        repo = self.user.create_repo_from_template("hello-world-docker-action-new", template_repo, description=description, private=private)
+        self.assertEqual(repo.description, description)
+        self.assertEqual(repo.private, private)
```
Comment (s-t-e-v-e-n-k) [inline tests/AuthenticatedUser.py]: ``self.assertTrue(repo.private)``

File: tests/Organization.py [hunk 3]
```diff
@@ -329,6 +329,22 @@ def testCreateFork(self):
...
+        self.assertEqual(repo.is_template, False)
+
+    def testCreateRepoFromTemplateWithAllArguments(self):
+        template_repo = self.g.get_repo("actions/hello-world-docker-action")
+
+        description = "My repo from template"
+        private = True
+        repo = self.org.create_repo_from_template("hello-world-docker-action-new", template_repo, description=description, private=private)
+        self.assertEqual(repo.description, description)
+        self.assertEqual(repo.private, private)
```
Comment (s-t-e-v-e-n-k) [inline tests/Organization.py]: As above

File: tests/Repository.py [hunk 4]
```diff
@@ -80,6 +80,7 @@ def testAttributes(self):
         self.assertEqual(self.repo.homepage, "http://vincent-jacques.net/PyGithub")
         self.assertEqual(self.repo.html_url, "https://github.com/jacquev6/PyGithub")
         self.assertEqual(self.repo.id, 3544490)
+        self.assertEqual(self.repo.is_template, None)
```
Comment (s-t-e-v-e-n-k) [inline tests/Repository.py]: ``self.assertIs(self.repo.is_template, None)``

File: github/Organization.py [hunk 5]
```diff
@@ -357,6 +357,47 @@ def create_fork(self, repo):
...
+        assert isinstance(repo, github.Repository.Repository), repo
+        assert description is github.GithubObject.NotSet or isinstance(
+            description, str
+        ), description
+        assert private is github.GithubObject.NotSet or isinstance(
+            private, bool
+        ), private
+        post_parameters = {
+            "name": name,
+            "owner": self.login,
```
Comment (s-t-e-v-e-n-k) [inline github/Organization.py]: I wonder if we should use self._identity here.

File: scripts/add_attribute.py (hunk 1 above)
Comment (s-t-e-v-e-n-k) [inline scripts/add_attribute.py]: This should be a separate PR, but this looks like confusing str and bytes before writing.

File: github/Organization.py [hunk 6]
```diff
@@ -357,6 +357,47 @@ def create_fork(self, repo):
             self._requester, headers, data, completed=True
         )
 
+    def create_repo_from_template(
```
Comment (s-t-e-v-e-n-k) [inline github/Organization.py]: Perhaps we should abstract this out to a mix-in, since it's identical...

**Comment by s-t-e-v-e-n-k** [comment]
I love the direction, but I think we need to have a think about how we implement this before we land it.

**Comment by sfdye** [comment]
# [Codecov](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=h1) Report
> Merging [#1395](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=desc) into [master](https://codecov.io/gh/PyGithub/PyGithub/commit/e5ce4912f5e96107eeb2670b8d29c173f82362f8&el=desc) will **increase** coverage by `0.14%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master    #1395      +/-   ##
==========================================
+ Coverage   98.20%   98.35%   +0.14%     
==========================================
  Files         188      188              
  Lines       14071    14140      +69     
==========================================
+ Hits        13819    13907      +88     
+ Misses        252      233      -19     
```


| [Impacted Files](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [tests/Framework.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvRnJhbWV3b3JrLnB5) | `96.00% <0.00%> (-0.02%)` | :arrow_down: |
| [tests/Logging\_.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvTG9nZ2luZ18ucHk=) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [tests/NamedUser.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvTmFtZWRVc2VyLnB5) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [tests/GitRelease.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvR2l0UmVsZWFzZS5weQ==) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [tests/Repository.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvUmVwb3NpdG9yeS5weQ==) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [tests/Persistence.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvUGVyc2lzdGVuY2UucHk=) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [tests/RateLimiting.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvUmF0ZUxpbWl0aW5nLnB5) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [github/Repository.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-Z2l0aHViL1JlcG9zaXRvcnkucHk=) | `97.29% <0.00%> (+0.74%)` | :arrow_up: |
| [tests/AllTests.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-dGVzdHMvQWxsVGVzdHMucHk=) | `100.00% <0.00%> (+1.07%)` | :arrow_up: |
| [github/Rate.py](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree#diff-Z2l0aHViL1JhdGUucHk=) | `100.00% <0.00%> (+4.16%)` | :arrow_up: |
| ... and [1 more](https://codecov.io/gh/PyGithub/PyGithub/pull/1395/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=footer). Last update [e5ce491...1aac6e9](https://codecov.io/gh/PyGithub/PyGithub/pull/1395?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: scripts/add_attribute.py (hunk 1 above)
Comment (isouza-daitan) [reply scripts/add_attribute.py]: I'll move this to a different PR, I think the real problem is that the file is opened in binary mode (`wb`)

**Comment by isouza-daitan** [comment]
@s-t-e-v-e-n-k thanks for the review!

The `create_fork` and `create_repo_from_template` methods are different than the other methods in the class because they are Repository endpoints, instead of Organization or User endpoints. Maybe we should move them to `Repository`? I think it would be harder to use but would at least match the API endpoint.

We could even keep the Org and AuthUser methods which would call the Repo method and set the appropriate owner. Thoughts?

File: github/Organization.py (hunk 5 above)
Comment (isouza-daitan) [reply github/Organization.py]: neither `Organization` nor `AuthenticatedUser` have an `_identity` property

**Comment by isouza-daitan** [comment]
@s-t-e-v-e-n-k thoughts on my comment above?

**Comment by s-t-e-v-e-n-k** [comment]
@isouza-daitan Sorry for the late reply. From a correctness perspective Repository is probably the right place, but given our architecture it's hard to require a user grab a repository object just to create a different repo, potentially in a different namespace entirely.

I'm still against the code duplication -- how about one method on AuthenicatedUser that also takes owner as NotSet. If owner is NotSet, use self.login, otherwise assert it's an Organization (or an id). What do you think?

**Comment by isouza-daitan** [comment]
@s-t-e-v-e-n-k won't we have auth problems if we try to create a new repo in an org using an AuthenticatedUser instance?

I agree getting a Repository to create a new repo is not great. 
How about a new `RepositoryActions` module that has:
```
@staticmethod
def create_repo_from_template(
            new_repo_owner,  # AuthenticatedUser or Organization
            new_repo_name,
            template_full_name,  # my-org/my-template-name
            description=github.GithubObject.NotSet,
            private=github.GithubObject.NotSet,
    ):
```
and we call `RepositoryActions.create_repo_from_template` from `AuthenticatedUser` and `Organization`

**Comment by sfdye** [comment]
1. Could you fix the python 3.6 tests?
2. I still see travis check in the PR, can you rebase against master?

**Comment by isacssouza** [comment]
@sfdye done, sorry for the delay.

**Comment by rue-nsilverman** [comment]
Its also worth adding that the `include_all_branches` parameter now exists on the create repo from template API. This would be hugely beneficial to add in to this function.

**Comment by simkimsia** [comment]
I kinda need this. Can solve the conflict? @isouza-daitan

**Comment by prein** [comment]
Please reopen and consider merging #2090
